proto/ibc/applications/transfer/v2/packet.proto

```diff
@@ -19,3 +21,27 @@ message FungibleTokenPacketData {
   // optional memo
   string memo = 5;
 }
+
+// FungibleTokenPacketDataV2 defines a struct for the packet payload
+// See FungibleTokenPacketDataV2 spec:
+// https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#data-structures
+message FungibleTokenPacketDataV2 {
+  // the tokens to be transferred
+  repeated Token tokens = 1 [(gogoproto.nullable) = false];
+  // the sender address
+  string sender = 2;
+  // the recipient address on the destination chain
+  string receiver = 3;
+  // optional memo
+  string memo = 4;
+}
+
+// Token defines a struct which represents a token to be transferred.
+message Token {
+  // the base token denomination to be transferred
+  string denom = 1;
+  // the token amount to be transferred
+  string amount = 2;
+  // the trace of the token
+  repeated string trace = 3;
```

Comment:
maybe we should add godoc on what this means? ie is each string a port/channel pair or isolated port and isolated channel?

Reply:
is there reason why we're not using something like `repeated Trace trace` where:

```proto
message Trace {
    string port_id = 1;
    string channel_id = 2;
}
```

I'm not sure if this was explored, just thought I'd mention here!

Reply:
I don't think it was explored. I had the same curiosity as it removes string parsing based on "/" (for v2 packets forward)

Reply:
what do you think about creating an issue for this? happy to defer til later but could be a nice improvement if we can remove string parsing

Reply:
I think I'm in favour after digging through the code some more. I think only going part way to typed information about the ibc token still causes confusion.

v1 approach:
ibc token: "portid/channelid/portid/channelid/denom", amount

v2 approach:
ibc token: denom, amount, "portid/channelid", "portid, channelid"

For backwards compatibility we still need to parse from v1 and reformat as v1, which means that we have logic to go from:
[]string{"portid", "channelid", "portid", "channelid"}
to:
[]string{"portid/channelid", "portid/channelid"}
and back

I find this confusing, as it is hard to know if `[]string` is single items or multiple. I highlighted in this issue how many extra parsing steps we do https://github.com/cosmos/ibc-go/issues/6221

I think with typed trace:
if recv v1
   parse the path string into token type

if send v1
   glue token type back into path string
   
otherwise always store fully typed trace info

Reply:
https://github.com/cosmos/ibc-go/issues/6362